File: CousinsInBT_DFS.py
# TC: O(N) where N is the total number of nodes of the tree
# SC: O(H) where H is the size of the recursive stack or the height of the binary tree.

class Solution:
    def isCousins(self, root: TreeNode, x: int, y: int) -> bool:
        self.x_depth = 0
        self.y_depth = 0
        
        self.x_parent = None
        self.y_parent = None
        
        def dfs(root, x, y, depth, parent):
            if not root:
                return
            
            if root.val == x:
                self.x_depth = depth
                self.x_parent = parent
            if root.val == y:
                self.y_depth = depth
                self.y_parent = parent
            
            dfs(root.left, x, y, depth + 1, root)
            dfs(root.right, x, y, depth + 1, root)
            
        dfs(root, x, y, 0, None)
        if self.x_depth == self.y_depth and self.x_parent != self.y_parent:
            return True
        return False
    

File: CousinsinBTDFS.java
//Time Complexity - O(n)
//Space Complexity - O(h)

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int xdepth; int ydepth;
    TreeNode xParent; TreeNode yParent;
  
    public boolean isCousins(TreeNode root, int x, int y) {
      if(root == null || root.val == x || root.val == y) return false;
      dfs(root,null,0,x,y);
      return xParent != yParent && xdepth == ydepth;
        
    }
  
    public void dfs(TreeNode root, TreeNode parent, int depth, int x, int y) {
      //base
      if(root == null) return;
      
      //logic
      //if the value x and y found then record parent and level
      if(root.val == x) {
        xdepth = depth;
        xParent = parent;
      }
      if(root.val == y) {
        ydepth = depth;
        yParent = parent;
      }
      //if not found continue search in left subtree and then right
      dfs(root.left,root, depth+1, x,y);
      dfs(root.right,root, depth+1, x,y);
      
    }
}